File: frontend/src/shared/lib/legal-entity.ts
const LEGAL_ENTITY_REPLACEMENTS: Array<[string, string]> = [
  ["ОБЩЕСТВО С ОГРАНИЧЕННОЙ ОТВЕТСТВЕННОСТЬЮ", "ООО"],
  ["ПУБЛИЧНОЕ АКЦИОНЕРНОЕ ОБЩЕСТВО", "ПАО"],
  ["НЕПУБЛИЧНОЕ АКЦИОНЕРНОЕ ОБЩЕСТВО", "НАО"],
  ["АКЦИОНЕРНОЕ ОБЩЕСТВО", "АО"],
  ["ИНДИВИДУАЛЬНЫЙ ПРЕДПРИНИМАТЕЛЬ", "ИП"],
  ["ФЕДЕРАЛЬНОЕ ГОСУДАРСТВЕННОЕ БЮДЖЕТНОЕ ОБРАЗОВАТЕЛЬНОЕ УЧРЕЖДЕНИЕ", "ФГБОУ"],
  ["ФЕДЕРАЛЬНОЕ ГОСУДАРСТВЕННОЕ БЮДЖЕТНОЕ УЧРЕЖДЕНИЕ", "ФГБУ"],
  ["ГОСУДАРСТВЕННОЕ БЮДЖЕТНОЕ УЧРЕЖДЕНИЕ", "ГБУ"],
  ["МУНИЦИПАЛЬНОЕ БЮДЖЕТНОЕ УЧРЕЖДЕНИЕ", "МБУ"],
];

function normalizeLegalEntityName(value: string) {
  return value
    .replace(/[\u00a0\u2000-\u200b\u202f\u205f\u3000]/g, " ")
    .replace(/[«»]/g, "\"")
    .trim()
    .replace(/\s+/g, " ");
}

export function abbreviateLegalEntityName(value: string) {
  const normalizedValue = normalizeLegalEntityName(value);
  const upperValue = normalizedValue.toUpperCase();

  for (const [fullName, abbreviation] of LEGAL_ENTITY_REPLACEMENTS) {
    const matchIndex = upperValue.indexOf(fullName);

    if (matchIndex === -1) {
      continue;
    }

    const before = normalizedValue.slice(0, matchIndex);
    const after = normalizedValue.slice(matchIndex + fullName.length).trimStart();
    const parts = [before.trimEnd(), abbreviation, after].filter(Boolean);

    return parts.join(" ").replace(/\s+/g, " ").trim();
  }

  return normalizedValue;
}

File: backend/src/utils/legal_entity.py
LEGAL_ENTITY_REPLACEMENTS: list[tuple[str, str]] = [
    ("ОБЩЕСТВО С ОГРАНИЧЕННОЙ ОТВЕТСТВЕННОСТЬЮ", "ООО"),
    ("ПУБЛИЧНОЕ АКЦИОНЕРНОЕ ОБЩЕСТВО", "ПАО"),
    ("НЕПУБЛИЧНОЕ АКЦИОНЕРНОЕ ОБЩЕСТВО", "НАО"),
    ("АКЦИОНЕРНОЕ ОБЩЕСТВО", "АО"),
    ("ИНДИВИДУАЛЬНЫЙ ПРЕДПРИНИМАТЕЛЬ", "ИП"),
    ("ФЕДЕРАЛЬНОЕ ГОСУДАРСТВЕННОЕ БЮДЖЕТНОЕ ОБРАЗОВАТЕЛЬНОЕ УЧРЕЖДЕНИЕ", "ФГБОУ"),
    ("ФЕДЕРАЛЬНОЕ ГОСУДАРСТВЕННОЕ БЮДЖЕТНОЕ УЧРЕЖДЕНИЕ", "ФГБУ"),
    ("ГОСУДАРСТВЕННОЕ БЮДЖЕТНОЕ УЧРЕЖДЕНИЕ", "ГБУ"),
    ("МУНИЦИПАЛЬНОЕ БЮДЖЕТНОЕ УЧРЕЖДЕНИЕ", "МБУ"),
]


def normalize_legal_entity_name(value: str) -> str:
    normalized_value = " ".join(value.replace("\u00A0", " ").strip().split())
    upper_value = normalized_value.upper()

    for full_name, abbreviation in LEGAL_ENTITY_REPLACEMENTS:
        match_index = upper_value.find(full_name)
        if match_index == -1:
            continue

        before = normalized_value[:match_index].strip()
        after = normalized_value[match_index + len(full_name):].strip()
        parts = [part for part in (before, abbreviation, after) if part]
        return " ".join(parts)

    return normalized_value
